execution/scan/matrix_selector.go

```diff
@@ -123,14 +118,32 @@ func (o *matrixSelector) Next(ctx context.Context) ([]model.StepVector, error) {
 			}
 			maxt := seriesTs - o.offset
 			mint := maxt - o.selectRange
-
 			rangePoints := selectPoints(series.samples, mint, maxt, o.scanners[i].previousPoints)
-			result := o.call(series.labels, rangePoints, seriesTs, o.selectRange)
-			if result.Point != InvalidSample.Point {
-				vectors[currStep].T = result.T
-				vectors[currStep].Samples = append(vectors[currStep].Samples, result.V)
-				vectors[currStep].SampleIDs = append(vectors[currStep].SampleIDs, series.signature)
+
+			if o.call != nil {
+				if len(vectors) <= currStep {
+					vectors = append(vectors, o.vectorPool.GetStepVector(seriesTs))
+				}
+
+				// TODO(saswatamcode): Handle multi-arg functions for matrixSelectors via injectable.
+				result := o.call(function.FunctionArgs{
+					Labels:      series.labels,
+					Points:      rangePoints,
+					StepTime:    seriesTs,
+					SelectRange: o.selectRange,
+				})
+
+				if result.Point != function.InvalidSample.Point {
+					vectors[currStep].T = result.T
+					vectors[currStep].Samples = append(vectors[currStep].Samples, result.V)
+					vectors[currStep].SampleIDs = append(vectors[currStep].SampleIDs, series.signature)
+				}
+			} else {
+				// TODO(saswatamcode): Range vector result might need new operator.
+				// https://github.com/thanos-community/promql-engine/issues/39
+				return nil, parse.ErrNotImplemented
```

Comment:
This error needs to be returned when the operator is created.

Reply:
Moved the error